Relative paths with double dot at the beginning (..\) fail in reports

PathUtils::sanitize in sonar-plugin-api is described as "Normalize path and replace file separators by forward slash", but the only thing it does is call `return FilenameUtils.normalize(path, true);` (https://github.com/SonarSource/sonar-plugin-api/blob/master/plugin-api/src/main/java/org/sonar/api/utils/PathUtils.java#L42). https://commons.apache.org/proper/commons-io/apidocs/org/apache/commons/io/FilenameUtils.html#normalize(java.lang.String,boolean) explains that

> A double dot will cause that path segment and the one before to be removed. If the double dot has no parent path segment to work with, null is returned. 

So the CxxReportLocation [constructor](https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/cxx-squid/src/main/java/org/sonar/cxx/utils/CxxReportLocation.java#L39) may return something with a line and a column, but a "null" file, if the "file" it's given is something like "../dir/file.c".

In https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxIssuesReportSensor.java#L164 only the file is checked. So, probably I should not have those `../` paths in my build log to start with, but if somebody happens to have them he ends up with issues being listed in SonarQube with no location. Usually, when a file is not "known", the issue is simply ignored (https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxReportSensor.java#L84), so there is a strange discrepancy here.

The issue can be reproduced with sonar-cxx 2.1.2 making `-Dsonar.cxx.gcc.reportPaths=` point to a file with something like `../file.c:1:1: warning: message [-Wdiv-by-zero]` (supposing `-Wdiv-by-zero` is in the quality profile).

--- a/cxx-squid/src/main/java/org/sonar/cxx/utils/CxxReportLocation.java
+++ b/cxx-squid/src/main/java/org/sonar/cxx/utils/CxxReportLocation.java
@@ -19,9 +19,9 @@
  */
 package org.sonar.cxx.utils;
 
-import java.nio.file.Paths;
 import java.util.Objects;
 import javax.annotation.Nullable;
+import org.apache.commons.io.FilenameUtils;
 import org.sonar.api.utils.PathUtils;
 
 /**
@@ -42,10 +42,13 @@ public class CxxReportLocation {
     // FilenameUtils.normalize internally, relative paths starting with a double dot will cause that path segment
     // and the one before to be removed. If the double dot has no parent path segment to work with, null is returned.
     // null would mean 'project issue' which is wrong in this context. To avoid this we extract the filename to
-    // generate at least a meningful error message.
+    // generate at least a meningful error message (#2747).
     var normalized = PathUtils.sanitize(file);
     if (normalized == null && (file != null && !file.isBlank())) {
-      normalized = Paths.get(file).getFileName().toString();
+
+      // use FilenameUtils.getName because this works on Windows and Linux also if
+      // report is generated on the one and consumed on the other
+      normalized = FilenameUtils.getName(file);
     }
 
     this.file = normalized;

--- a/cxx-squid/src/main/java/org/sonar/cxx/utils/CxxReportLocation.java
+++ b/cxx-squid/src/main/java/org/sonar/cxx/utils/CxxReportLocation.java
@@ -19,6 +19,7 @@
  */
 package org.sonar.cxx.utils;
 
+import java.nio.file.Paths;
 import java.util.Objects;
 import javax.annotation.Nullable;
 import org.sonar.api.utils.PathUtils;
@@ -35,8 +36,19 @@ public class CxxReportLocation {
 
   public CxxReportLocation(@Nullable String file, @Nullable String line, @Nullable String column, String info) {
     super();
-    // normalize file, removing double and single dot path steps => avoids duplicates
-    this.file = PathUtils.sanitize(file);
+
+    // Normalize file using separators in UNIX format, removing double and single dot path steps. This is to avoid
+    // duplicates in the issue containers because they are using the file as key. PathUtils.sanitize uses
+    // FilenameUtils.normalize internally, relative paths starting with a double dot will cause that path segment
+    // and the one before to be removed. If the double dot has no parent path segment to work with, null is returned.
+    // null would mean 'project issue' which is wrong in this context. To avoid this we extract the filename to
+    // generate at least a meningful error message.
+    var normalized = PathUtils.sanitize(file);
+    if (normalized == null && (file != null && !file.isBlank())) {
+      normalized = Paths.get(file).getFileName().toString();
+    }
+
+    this.file = normalized;
     this.line = line;
     this.column = column;
     this.info = info;

--- a/cxx-squid/src/test/java/org/sonar/cxx/utils/CxxReportLocationTest.java
+++ b/cxx-squid/src/test/java/org/sonar/cxx/utils/CxxReportLocationTest.java
@@ -1,0 +1,78 @@
+/*
+ * C++ Community Plugin (cxx plugin)
+ * Copyright (C) 2010-2023 SonarOpenCommunity
+ * http://github.com/SonarOpenCommunity/sonar-cxx
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.cxx.utils;
+
+import static org.assertj.core.api.Assertions.*;
+import org.junit.jupiter.api.Test;
+
+public class CxxReportLocationTest {
+
+  @Test
+  void testConstructor() {
+    var loc0 = new CxxReportLocation(null, null, null, "info");
+    assertThat(loc0.getFile()).isNull();
+    assertThat(loc0.getLine()).isNull();
+    assertThat(loc0.getColumn()).isNull();
+    assertThat(loc0.getInfo()).isEqualTo("info");
+    assertThat(loc0.toString()).isEqualTo("CxxReportLocation [file=null, line=null, column=null, info=info]");
+
+    var loc1 = new CxxReportLocation("file", "line", "column", "info");
+    assertThat(loc1.getFile()).isEqualTo("file");
+    assertThat(loc1.getLine()).isEqualTo("line");
+    assertThat(loc1.getColumn()).isEqualTo("column");
+    assertThat(loc1.getInfo()).isEqualTo("info");
+    assertThat(loc1.toString()).isEqualTo("CxxReportLocation [file=file, line=line, column=column, info=info]");
+  }
+
+  @Test
+  void testPathSanitize() {
+    var loc = new CxxReportLocation("file", null, null, "");
+
+    loc = new CxxReportLocation("/dir/File", null, null, "");
+    assertThat(loc.getFile()).isEqualTo("/dir/File");
+    loc = new CxxReportLocation("/dir/./File", null, null, "");
+    assertThat(loc.getFile()).isEqualTo("/dir/File");
+    loc = new CxxReportLocation("/dir/../File", null, null, "");
+    assertThat(loc.getFile()).isEqualTo("/File");
+    loc = new CxxReportLocation("dir/File", null, null, "");
+    assertThat(loc.getFile()).isEqualTo("dir/File");
+    loc = new CxxReportLocation("./dir/File", null, null, "");
+    assertThat(loc.getFile()).isEqualTo("dir/File");
+    loc = new CxxReportLocation("../dir/File", null, null, "");
+    assertThat(loc.getFile()).isEqualTo("File");
+    loc = new CxxReportLocation("../../File", null, null, "");
+    assertThat(loc.getFile()).isEqualTo("File");
+
+    loc = new CxxReportLocation("c:\\dir\\file.ext", null, null, "");
+    assertThat(loc.getFile()).isEqualTo("c:/dir/file.ext");
+    loc = new CxxReportLocation("C:\\dir\\.\\file.ext", null, null, "");
+    assertThat(loc.getFile()).isEqualTo("C:/dir/file.ext");
+    loc = new CxxReportLocation("c:\\dir\\..\\file.ext", null, null, "");
+    assertThat(loc.getFile()).isEqualTo("c:/file.ext");
+    loc = new CxxReportLocation("dir\\file.ext", null, null, "");
+    assertThat(loc.getFile()).isEqualTo("dir/file.ext");
+    loc = new CxxReportLocation(".\\dir\\file.ext", null, null, "");
+    assertThat(loc.getFile()).isEqualTo("dir/file.ext");
+    loc = new CxxReportLocation("..\\dir\\file.ext", null, null, "");
+    assertThat(loc.getFile()).isEqualTo("file.ext");
+    loc = new CxxReportLocation("..\\..\\file.ext", null, null, "");
+    assertThat(loc.getFile()).isEqualTo("file.ext");
+  }
+}

